File: app/src/core/matching.ts
/** Frekans eşleştirme teknikleri — engine/aurora/matching.py ile aynı mantık. */
import { REFERENCE_RATIOS, cents, octaveReduce } from "./math";

export interface CentMatch { cents: number; match: boolean; octaveEquivalent: boolean }
export interface RationalMatch { ratio: string; value: number; errorCents: number; match: boolean; referenceLabel: string | null; complexity: number }
export interface HarmonicMatch { direct: { kind: "f1_of_f2" | "f2_of_f1"; k: number } | null; common: { p: number; q: number; commonHz: number; errorCents: number } | null; match: boolean }

/** 1–2. Cent uzaklığı; oktav eşdeğerliği ile çevrimsel mesafe */
export const centMatch = (f1: number, f2: number, tol = 8, octaveEquivalent = false): CentMatch => {
  let c = octaveEquivalent ? cents(octaveReduce(f1), octaveReduce(f2)) : cents(f1, f2);
  if (octaveEquivalent) c = Math.min(c, 1200 - c);
  return { cents: c, match: c <= tol, octaveEquivalent };
};

/** Sürekli kesir ile payda sınırlı en iyi rasyonel yaklaşım (Fraction.limit_denominator eşdeğeri). */
export const limitDenominator = (x: number, maxDen = 16): [number, number] => {
  let best: [number, number] = [Math.round(x), 1];
  let bestErr = Math.abs(x - best[0]);
  for (let d = 1; d <= maxDen; d++) {
    const n = Math.round(x * d);
    const err = Math.abs(x - n / d);
    if (n > 0 && err < bestErr - 1e-12) {
      best = [n, d];
      bestErr = err;
    }
  }
  const g = gcd(best[0], best[1]);
  return [best[0] / g, best[1] / g];
};

const gcd = (a: number, b: number): number => (b === 0 ? a : gcd(b, a % b));

/** 3. Rasyonel oran eşleştirme */
export const rationalMatch = (f1: number, f2: number, maxDen = 16, tol = 8): RationalMatch => {
  const x = f1 / f2;
  const [n, d] = limitDenominator(x, maxDen);
  const err = cents(x, n / d);
  const label =
    Object.entries(REFERENCE_RATIOS).find(([, [rn, rd]]) => (rn === n && rd === d) || (rn === d && rd === n))?.[0] ?? null;
  return { ratio: `${n}:${d}`, value: n / d, errorCents: err, match: err <= tol, referenceLabel: label, complexity: n + d };
};

/** 4. Harmonik eşleştirme: f1 ≈ k·f2 / ortak harmonik p·f1 ≈ q·f2 */
export const harmonicMatch = (f1: number, f2: number, maxH = 16, tol = 8): HarmonicMatch => {
  let direct: HarmonicMatch["direct"] = null;
  for (let k = 1; k <= maxH && !direct; k++) {
    if (cents(f1, k * f2) <= tol) direct = { kind: "f1_of_f2", k };
    else if (cents(f2, k * f1) <= tol) direct = { kind: "f2_of_f1", k };
  }
  let common: HarmonicMatch["common"] = null;
  for (let p = 1; p <= maxH; p++)
    for (let q = 1; q <= maxH; q++) {
      const err = cents(p * f1, q * f2);
      if (err <= tol && (!common || p + q < common.p + common.q)) common = { p, q, commonHz: p * f1, errorCents: err };
    }
  return { direct, common, match: Boolean(direct || common) };
};

/** 5. Spektral tepe ↔ hedef eşleştirme */
export const spectralMatch = (peaks: Array<[number, number]>, targets: number[], tol = 15) =>
  targets.map((t) => {
    const cand = peaks.filter(([p]) => p > 0).map(([p, m]) => ({ c: cents(p, t), p, m })).sort((a, b) => a.c - b.c)[0];
    return cand ? { target: t, peakHz: cand.p, magnitude: cand.m, errorCents: cand.c, match: cand.c <= tol } : { target: t, match: false };
  });

/** 6. DTW — iki frekans dizisi arasında hizalanmış cent uzaklığı (uzunluk normalize) */
export const dtwDistance = (a: number[], b: number[], metric: "cents" | "abs" = "cents"): number => {
  if (!a.length || !b.length) return Infinity;
  const d = (x: number, y: number) => {
    if (metric === "abs") return Math.abs(x - y);
    if (x <= 0 || y <= 0) return x === y ? 0 : 2400;
    return Math.abs(1200 * Math.log2(x / y));
  };
  const D = Array.from({ length: a.length + 1 }, () => new Array<number>(b.length + 1).fill(Infinity));
  D[0][0] = 0;
  for (let i = 1; i <= a.length; i++)
    for (let j = 1; j <= b.length; j++)
      D[i][j] = d(a[i - 1], b[j - 1]) + Math.min(D[i - 1][j], D[i][j - 1], D[i - 1][j - 1]);
  return D[a.length][b.length] / (a.length + b.length);
};

/** Sekans 'grameri': ardışık geçişlerin işaretli cent değerleri */
export const intervalProfile = (seq: number[]): number[] =>
  seq.slice(1).map((b, i) => (seq[i] > 0 && b > 0 ? 1200 * Math.log2(b / seq[i]) : 0));

export const sequenceMatch = (a: number[], b: number[], tol = 50) => {
  const dtw = dtwDistance(a, b);
  const pa = intervalProfile(a).map((x) => 1000 + x);
  const pb = intervalProfile(b).map((x) => 1000 + x);
  const prof = pa.length && pb.length ? dtwDistance(pa, pb, "abs") : Infinity;
  return { dtwCents: dtw, intervalProfileDistance: prof, match: dtw <= tol, transpositionInvariantMatch: prof <= tol };
};

/** 7. Parmak izi benzerliği (0–1) — enerji benzerliği DEĞİL, geometrik benzerlik */
export const vectorSimilarity = (va: number[], vb: number[]): number => {
  const dist = Math.sqrt(va.reduce((s, x, i) => s + (x - vb[i]) ** 2, 0));
  return Math.max(0, 1 - dist / Math.sqrt(va.length));
};

File: engine/aurora/matching.py
"""Matching Engine — frekans eşleştirme teknikleri.

1. Cent uzaklığı (perde algısı)          cents = 1200·log2(f1/f2)
2. Oktav eşdeğerliği                      f ~ f·2^k  → octave class
3. Rasyonel oran eşleştirme               f1/f2 ≈ n/d (Stern–Brocot / sürekli kesir)
4. Harmonik eşleştirme                    f1 = k·f2 veya ortak harmonik (LCM benzeri)
5. Spektral tepe eşleştirme               ölçülen tepe ↔ hedef Hz (tolerans cent)
6. Sekans eşleştirme (DTW)                iki frekans dizisi arasında hizalanmış uzaklık
7. Desen parmak izi benzerliği            chladni.similarity / mask IoU
8. Kümeleme                               feature vector → k-means (numpy)
"""
from __future__ import annotations

from fractions import Fraction
from math import gcd, log2

import numpy as np

from .math_analysis import REFERENCE_RATIOS, cents, octave_reduce


# 1–2 ---------------------------------------------------------------------

def cent_match(f1: float, f2: float, tolerance_cents: float = 8.0, octave_equivalent: bool = False) -> dict:
    a, b = (octave_reduce(f1), octave_reduce(f2)) if octave_equivalent else (f1, f2)
    c = cents(a, b)
    if octave_equivalent:
        c = min(c, 1200 - c)
    return {"f1": f1, "f2": f2, "cents": round(c, 3), "match": c <= tolerance_cents,
            "octave_equivalent": octave_equivalent}


def octave_class(f: float, precision: int = 4) -> float:
    return round(octave_reduce(f), precision)


# 3 -------------------------------------------------------------------------

def rational_match(f1: float, f2: float, max_denominator: int = 16, tolerance_cents: float = 8.0) -> dict:
    """f1/f2 oranına en yakın basit kesir ve cent hatası. Sürekli kesir (Fraction.limit_denominator)."""
    if f1 <= 0 or f2 <= 0:
        raise ValueError("frequencies must be positive")
    x = f1 / f2
    fr = Fraction(x).limit_denominator(max_denominator)
    err = cents(x, float(fr)) if fr > 0 else float("inf")
    label = next((k for k, (n, d) in REFERENCE_RATIOS.items()
                  if Fraction(n, d) == fr or Fraction(d, n) == fr), None)
    return {"ratio": f"{fr.numerator}:{fr.denominator}", "value": float(fr), "error_cents": round(err, 3),
            "match": err <= tolerance_cents, "reference_label": label, "complexity": fr.numerator + fr.denominator}


# 4 -------------------------------------------------------------------------

def harmonic_match(f1: float, f2: float, max_harmonic: int = 16, tolerance_cents: float = 8.0) -> dict:
    """f1 ≈ k·f2 veya f2 ≈ k·f1; ayrıca en küçük ortak harmonik (p·f1 ≈ q·f2)."""
    best = None
    for p in range(1, max_harmonic + 1):
        for q in range(1, max_harmonic + 1):
            err = cents(p * f1, q * f2)
            if err <= tolerance_cents and (best is None or p + q < best["p"] + best["q"]):
                best = {"p": p, "q": q, "common_hz": round(p * f1, 3), "error_cents": round(err, 3)}
    direct = None
    for k in range(1, max_harmonic + 1):
        if cents(f1, k * f2) <= tolerance_cents:
            direct = {"f1_is_harmonic_of_f2": k}
            break
        if cents(f2, k * f1) <= tolerance_cents:
            direct = {"f2_is_harmonic_of_f1": k}
            break
    return {"direct": direct, "common_harmonic": best, "match": bool(direct or best)}


# 5 -------------------------------------------------------------------------

def spectral_match(peaks: list[tuple[float, float]], targets: list[float], tolerance_cents: float = 15.0) -> list[dict]:
    """Ölçülen (hz, mag) tepeleri hedef frekanslarla eşler; her hedef için en yakın tepe."""
    out = []
    for t in targets:
        cand = [(cents(p, t), p, m) for p, m in peaks if p > 0]
        if not cand:
            out.append({"target": t, "peak": None, "match": False})
            continue
        c, p, m = min(cand)
        out.append({"target": t, "peak_hz": round(p, 3), "magnitude": round(m, 4),
                    "error_cents": round(c, 3), "match": c <= tolerance_cents})
    return out


# 6 -------------------------------------------------------------------------

def dtw_distance(seq_a: list[float], seq_b: list[float], metric: str = "cents") -> float:
    """Dynamic Time Warping — farklı uzunluktaki iki frekans dizisi arasında hizalanmış uzaklık."""
    if not seq_a or not seq_b:
        return float("inf")
    def d(x: float, y: float) -> float:
        if metric == "cents":
            if x <= 0 or y <= 0:
                return 0.0 if x == y else 2400.0
            return abs(1200 * log2(x / y))
        return abs(x - y)
    n, m = len(seq_a), len(seq_b)
    D = np.full((n + 1, m + 1), np.inf)
    D[0, 0] = 0.0
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            D[i, j] = d(seq_a[i - 1], seq_b[j - 1]) + min(D[i - 1, j], D[i, j - 1], D[i - 1, j - 1])
    return float(D[n, m] / (n + m))


def interval_profile(seq: list[float]) -> list[float]:
    """Sekansın 'gramer'i: ardışık geçişlerin cent değerleri (işaretli)."""
    return [round(1200 * log2(b / a), 3) if a > 0 and b > 0 else 0.0 for a, b in zip(seq, seq[1:])]


def sequence_match(seq_a: list[float], seq_b: list[float], tolerance: float = 50.0) -> dict:
    dist = dtw_distance(seq_a, seq_b)
    pa, pb = interval_profile(seq_a), interval_profile(seq_b)
    prof = dtw_distance([1000 + x for x in pa], [1000 + x for x in pb], metric="abs") if pa and pb else float("inf")
    return {"dtw_cents": round(dist, 3), "interval_profile_distance": round(prof, 3),
            "match": dist <= tolerance, "transposition_invariant_match": prof <= tolerance}


# 8 -------------------------------------------------------------------------

def kmeans(X: np.ndarray, k: int = 3, iters: int = 50, seed: int = 0) -> tuple[np.ndarray, np.ndarray]:
    rng = np.random.default_rng(seed)
    X = np.asarray(X, dtype=float)
    C = X[rng.choice(len(X), size=min(k, len(X)), replace=False)]
    labels = np.zeros(len(X), dtype=int)
    for _ in range(iters):
        dist = ((X[:, None, :] - C[None, :, :]) ** 2).sum(axis=2)
        new = dist.argmin(axis=1)
        if np.array_equal(new, labels) and _ > 0:
            break
        labels = new
        for j in range(len(C)):
            if np.any(labels == j):
                C[j] = X[labels == j].mean(axis=0)
    return labels, C


def cluster_frequencies(freqs: list[float], k: int = 3, seed: int = 0) -> dict[int, list[float]]:
    from .discovery import feature_vector
    X = np.stack([feature_vector(f) for f in freqs])
    X = (X - X.mean(axis=0)) / (X.std(axis=0) + 1e-9)
    labels, _ = kmeans(X, k, seed=seed)
    out: dict[int, list[float]] = {}
    for f, l in zip(freqs, labels):
        out.setdefault(int(l), []).append(f)
    return out


def match_report(f1: float, f2: float) -> dict:
    return {"cents": cent_match(f1, f2), "cents_octave": cent_match(f1, f2, octave_equivalent=True),
            "rational": rational_match(f1, f2), "harmonic": harmonic_match(f1, f2)}
